z3/src/z3_solver.cpp

```diff
@@ -71,19 +130,64 @@ Term Z3Solver::get_selector(const Sort &s, std::string con,
 ;
 
 Term Z3Solver::make_term(int64_t i, const Sort &sort) const {
-	throw NotImplementedException(
-			"Term iteration not implemented for Z3 backend.");
+	SortKind sk = sort->get_sort_kind();
+	expr z_term = expr(ctx);
+	if (sk == INT) {
+		z_term = ctx.int_val(i);
+	} else if (sk == REAL) {
+		z_term = ctx.real_val(i);
+	} else if (sk == BV) {
+		z_term = ctx.bv_val(sort->get_width(), i);
+	} else {
+		string msg("Can't create value ");
+		msg += i;
+		msg += " with sort ";
+		msg += sort->to_string();
+		throw IncorrectUsageException(msg);
+	}
+
+	return std::make_shared < Z3Term > (z_term, ctx);
 }
 
 Term Z3Solver::make_term(const std::string val, const Sort &sort,
 		uint64_t base) const {
-	throw NotImplementedException(
-			"Term iteration not implemented for Z3 backend.");
+	expr z_term = expr(ctx);
+	SortKind sk = sort->get_sort_kind();
+
+	if (base != 10) {
+		throw NotImplementedException(
+			"Does not support base not equal to 10.");
+	}
+
+	if (sk == BV) {
+		z_term = ctx.bv_const(val.c_str(), sort->get_width());
+	} else if (sk == REAL) {
+		z_term = ctx.real_const(val.c_str());
+	} else if (sk == INT) {
+		z_term = ctx.int_const(val.c_str());
+	} else {
+		string msg("Can't create value ");
+		msg += val;
+		msg += " with sort ";
+		msg += sort->to_string();
+		throw IncorrectUsageException(msg);
+	}
+
+	return std::make_shared < Z3Term > (z_term, ctx);
 }
 
 Term Z3Solver::make_term(const Term &val, const Sort &sort) const {
-	throw NotImplementedException(
-			"Constant arrays not supported for Z3 backend.");
+
+	std::shared_ptr<Z3Term> zterm = std::static_pointer_cast < Z3Term > (val);
+	std::shared_ptr<Z3Sort> zsort = std::static_pointer_cast < Z3Sort > (sort);
+
+	if (zsort->is_function || zterm->is_function) {
+		throw IncorrectUsageException("Arrays in functions are not supported for Z3"); //is this true?
```

Comment:
Regarding the comment - that seems reasonable to me. I don't think any solver supports something like that. This function is called `make_term` so it's consistent with the rest of the functions, but is really just for creating constant arrays. E.g. an array where every element is set to a constant value like 0. Doesn't really make sense to combine it with functions (or that would at least be higher order logic). You could make the error message a bit more specific if you want. Something like: `Cannot create constant array with function element`.